internal/util/flowgraph/node.go

```diff
@@ -95,35 +90,37 @@ func (nodeCtx *nodeCtx) Start(ctx context.Context, wg *sync.WaitGroup) {
 			w := sync.WaitGroup{}
 			for i := 0; i < downstreamLength; i++ {
 				w.Add(1)
-				go nodeCtx.downstream[i].ReceiveMsg(&w, res[i], nodeCtx.downstreamInputChanIdx[nodeCtx.downstream[i].node.Name()])
+				go nodeCtx.downstream[i].deliverMsg(&w, res[i], nodeCtx.downstreamInputChanIdx[nodeCtx.downstream[i].node.Name()])
 			}
 			w.Wait()
 		}
 	}
 }
 
+// Close handles cleanup logic and notify worker to quit
 func (nodeCtx *nodeCtx) Close() {
-	// data race with nodeCtx.ReceiveMsg { nodeCtx.inputChannels[inputChanIdx] <- msg }
-	//for _, channel := range nodeCtx.inputChannels {
-	//	close(channel)
-	//	log.Warn("close inputChannel")
-	//}
+	// close Node
 	nodeCtx.node.Close()
+	// notify worker
+	close(nodeCtx.closeCh)
```

Comment:
Do we want to close inputChannels? what

Reply:
Downstream node closes before current node is a possible situation and was the root cause of goroutine leakage. In `deliverMsg` we have a select to cancel send to channel operation to handle this case.